pkg/query-service/ee/usage/manager.go

```diff
@@ -0,0 +1,291 @@
+
+	for {
+		select {
+		case <-lm.done:
+			return
+		case <-collectionTicker.C:
+			lm.CollectAndStoreUsage(ctx)
+		case <-uploadTicker.C:
+			lm.UploadUsage(ctx)
+		}
+	}
+}
+
+type TableSize struct {
+	Table             string `ch:"table"`
+	DiskName          string `ch:"disk_name"`
+	Rows              uint64 `ch:"rows"`
+	UncompressedBytes uint64 `ch:"uncompressed_bytes"`
+}
+
+func (lm *Manager) CollectAndStoreUsage(ctx context.Context) error {
+	snap, err := lm.GetUsageFromClickHouse(ctx)
+	if err != nil {
+		return err
+	}
+
+	license, err := lm.licenseRepo.GetActiveLicense(ctx)
+	if err != nil {
+		return err
+	}
+
+	activationId, _ := uuid.Parse(license.ActivationId)
+	// TODO (nitya) : Add installation ID in the payload
+	payload := model.UsagePayload{
+		UsageBase: model.UsageBase{
+			ActivationId:      activationId,
+			FailedSyncRequest: 0,
+		},
+		Metrics:      *snap,
+		SnapshotDate: time.Now(),
+	}
+
+	_, err = lm.repo.InsertSnapshot(ctx, payload)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (lm *Manager) GetUsageFromClickHouse(ctx context.Context) (*model.UsageSnapshot, error) {
+	tableSizes := []TableSize{}
+	snap := model.UsageSnapshot{}
+
+	// get usage from clickhouse
+	query := `
+		SELECT
+			table,
+			disk_name,
+			sum(rows) as rows,
+			sum(data_uncompressed_bytes) AS uncompressed_bytes
+		FROM system.parts
+		WHERE active AND (database in ('signoz_logs', 'signoz_metrics', 'signoz_traces')) AND (table in ('logs','samples_v2', 'signoz_index_v2'))
+		GROUP BY
+			table,
+			disk_name
+		ORDER BY table
+	`
+	err := lm.repo.clickhouseConn.Select(ctx, &tableSizes, query)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, val := range tableSizes {
+		switch val.Table {
+		case "logs":
+			if val.DiskName == "default" {
+				snap.CurrentLogSizeBytes = val.UncompressedBytes
+			} else {
+				snap.CurrentLogSizeBytesColdStorage = val.UncompressedBytes
+			}
+		case "samples_v2":
+			if val.DiskName == "default" {
+				snap.CurrentTimeseriesCount = val.Rows
+			} else {
+				snap.CurrentTimeseriesCountColdStorage = val.Rows
+			}
+		case "signoz_index_v2":
+			if val.DiskName == "default" {
+				snap.CurrentSpansCount = val.Rows
+			} else {
+				snap.CurrentSpansCountColdStorage = val.Rows
+			}
+		}
+	}
+
+	return &snap, nil
+}
+
+func (lm *Manager) UploadUsage(ctx context.Context) error {
+	snapshots, err := lm.repo.GetSnapshotsNotSynced(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(snapshots) <= 0 {
+		zap.S().Info("no snapshots to upload, skipping.")
+		return nil
+	}
+
+	zap.S().Info("uploading snapshots")
+	for _, snap := range snapshots {
+		metricsBytes, err := encryption.Decrypt([]byte(snap.ActivationId.String()[:32]), []byte(snap.Snapshot))
+		if err != nil {
+			return err
+		}
+
+		metrics := model.UsageSnapshot{}
+		err = json.Unmarshal(metricsBytes, &metrics)
+		if err != nil {
+			return err
+		}
+
+		err = lm.UploadUsageWithExponentalBackOff(ctx, model.UsagePayload{
```

Comment:
Probably not a much concern but we could may be batch this upload and update part.

Reply:
yeah I did discuss this with @ankitnayan , since batching will require another level of standardisation w.r.t how we handle failues and responses, not tackling it now.